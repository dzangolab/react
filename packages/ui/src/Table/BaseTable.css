button {
  cursor: pointer;
}

table {
  border-collapse: collapse;
  border-spacing: 0;
  table-layout: fixed;
  width: 100%;
}

.sticky-header table {
  position: relative;
}

table caption {
  color: #333;
  font-size: 24px;
  font-weight: bold;
  margin-bottom: 16px;
  text-align: left;
}

table > thead > button {
  display: none;
}

tbody tr:nth-child(even) {
  background-color: #f2f2f2;
}

tbody tr:nth-child(odd) {
  background-color: #fff;
}

th,
td {
  border: 1px solid lightgray;
  padding: 8px;
  text-align: left;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

th > div > span {
  font-size: 1rem;
  font-weight: 600;
  width: max-content;
}

.sticky-header thead {
  position: sticky;
  top: 0;
}

.current-page-info {
  font-size: 18px;
  width: max-content;
}

.disable-select {
  align-items: center;
  cursor: pointer;
  display: flex;
  user-select: none;
}

.disable-select > button,
.filter-wrapper > button {
  background-color: transparent;
  border: 0;
  height: 1rem;
  opacity: 0;
  outline: 0;
  padding: 0;
  transition: visibility 0s linear 300ms, opacity 300ms;
  visibility: hidden;
  width: 1rem;
}

.disable-select:hover > button,
.disable-select:hover > .filter-wrapper > button {
  opacity: 1;
  transition: visibility 0s linear 0s, opacity 300ms;
  visibility: visible;
}

.disable-select > button,
.disable-select > button > img,
.filter-wrapper,
.filter-wrapper > button,
.filter-wrapper > button > img {
  height: 1rem;
  width: 1rem;
}

.filter-form-wrapper {
  background-color: #fff;
  border: solid 2px gray;
  margin-left: -40px;
  margin-top: 5px;
  padding: 8px;
  position: absolute;
  z-index: 1000;
}

.number-filter-form {
  display: flex;
  flex-direction: column;
}

.number-filter-form > *:not(:last-child) {
  margin-bottom: 4px;
}

.page-details {
  align-items: center;
  display: flex;
  flex-wrap: wrap;
}

.page-details > *:not(:last-child) {
  margin-right: 16px;
}

.page-controller {
  align-items: center;
  display: flex;
  flex-wrap: wrap;
}

.page-controller > *:not(:last-child) {
  margin-right: 8px;
}

.page-controller span {
  font-size: 18px;
  width: max-content;
}

.pagination-button {
  align-items: center;
  background-color: #fff;
  border: solid 1px gray;
  border-radius: 50%;
  display: flex;
  height: 32px;
  justify-content: center;
  transition: 0.3s;
  width: 32px;
}

.pagination-button[disabled] {
  background-color: #e7e7e7;
  cursor: not-allowed;
}

.pagination-button:not[disabled]:hover {
  background-color: rgb(222, 222, 222);
}

.pagination-button > img {
  height: 1rem;
  width: 1rem;
}

.pagination-container {
  align-items: center;
  display: flex;
}

.pagination-container > *:not(:last-child) {
  margin-right: 4px;
}

.table-info-container {
  align-items: center;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
}

.table-info-container > * {
  margin-top: 16px;
}

.table-info-container > *:not(:last-child) {
  margin-right: 16px;
}

.table-wrapper {
  display: block;
  max-width: 100%;
  overflow-x: scroll;
  overflow-y: hidden;
}

.sticky-header .table-wrapper {
  max-height: 60vh;
  overflow-y: scroll;
}

.hide-scrollbar .table-wrapper {
  -ms-overflow-style: none;
  scrollbar-width: none;
}

.hide-scrollbar .table-wrapper::-webkit-scrollbar {
  display: none;
}

table > thead > tr:first-child {
  display: none;
}

table > thead > tr:last-child {
  background-color: rgb(58, 58, 58);
  color: #fff;
}

@media screen and (max-width: 600px) {

  table > thead > tr:first-child {
    display: block;
    background-color: #fff;
  }

  table > thead > tr:first-child > th {
    border: 0;
    padding: 0;
  }

  table > thead > tr:first-child > th > button {
    all: unset;
    background-color: rgb(58, 58, 58);
    cursor: pointer;
    display: flex;
    width: 1.5rem;
    height: 1.5rem;
    margin-bottom: 0.5rem;
  }

  table > thead > tr:not(:first-child) {
    display: none;
  }

  table > thead.active > tr:last-child {
    display: flex;
    flex-direction: column;
  }

  table > thead > tr > th,
  table > tbody > tr > td {
    width: 100% !important; /* override column size from props */
  }

  table > tbody > tr {
    display: block;
    margin-bottom: var(--table-cards-gap, 0.6rem);
  }

  table td {
    border: 1px solid #222;
    border-bottom: none;
    display: block;
    font-size: var(--td-font-size, 0.8rem);
    text-align: right;
  }

  table td::before {
    content: attr(data-label);
    float: left;
    font-weight: bold;
    text-transform: uppercase;
  }

  table td:last-child {
    border-bottom: var(--td-border-bottom, 1px solid #222);
  }
}
