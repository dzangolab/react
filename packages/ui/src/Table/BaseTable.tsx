import { useContext } from "react";

import TableBody from "./TableBody";
import TableFooter from "./TableFooter";
import TableHeader from "./TableHeader";
import { TableContext } from "./TableProvider";

function BaseTable() {
  const { fixedHeader, hideScrollBar, title } = useContext(TableContext);

  const getTableWrapperStyle = () => {
    if (fixedHeader && hideScrollBar) return "sticky-header hide-scrollbar";
    if (fixedHeader && !hideScrollBar) return "sticky-header";
    if (!fixedHeader && hideScrollBar) return "hide-scrollbar";
    return "";
  };

  return (
    <div className={`table-wrapper ${getTableWrapperStyle()}`}>
      <div>
        <table>
          {title ? <caption>{title}</caption> : null}
          <TableHeader />
          <TableBody />
        </table>
      </div>

      <TableFooter />
    </div>
  );
}

export default BaseTable;
