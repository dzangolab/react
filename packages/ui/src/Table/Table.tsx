import {
  getCoreRowModel,
  useReactTable,
  getPaginationRowModel,
  SortingState,
  getSortedRowModel,
  ColumnFiltersState,
  getFilteredRowModel,
  VisibilityState,
  PaginationState,
  Updater,
} from "@tanstack/react-table";
import React, { useEffect, useMemo, useRef, useState } from "react";

import { getStorage } from "../utils";
import {
  DEFAULT_PAGE_INDEX,
  DEFAULT_PAGE_PER_OPTIONS,
  DEFAULT_PAGE_SIZE,
} from "./constants";
import { TableBody } from "./TableBody";
import { DataActionsMenu } from "./TableDataActions";
import { Table, TableFooter } from "./TableElements";
import { TableHeader } from "./TableHeader";
import { TableToolbar } from "./TableToolbar";
import {
  getRequestJSON,
  getParsedColumns,
  saveTableState,
  getSavedTableState,
} from "./utils";
import { Checkbox } from "../FormWidgets";
import LoadingIcon from "../LoadingIcon";
import { Pagination } from "../Pagination";

import type { PersistentTableState, TDataTableProperties } from "./types";
import type { ColumnDef } from "@tanstack/react-table";

const DataTable = <TData extends { id: string | number }>({
  className = "",
  columnActionBtnLabel: columnActionButtonLabel = "Columns",
  columns = [],
  customFormatters,
  data,
  dataActionsMenu,
  emptyTableMessage,
  enableRowSelection = false,
  fetchData,
  id,
  initialFilters = [],
  initialSorting = [],
  inputDebounceTime,
  isLoading = false,
  onRowSelectChange,
  paginated = true,
  paginationOptions,
  persistState = false,
  persistStateStorage = "localStorage",
  renderCustomPagination,
  renderSortIcons,
  renderTableFooterContent,
  renderToolbarItems,
  resetStateActionBtnLabel: resetStateActionButtonLabel = "Reset all",
  rowPerPage,
  rowPerPageOptions = DEFAULT_PAGE_PER_OPTIONS,
  showColumnsAction = false,
  showResetStateAction,
  title,
  totalRecords = 0,
  visibleColumns = [],
  ...tableOptions
}: TDataTableProperties<TData>) => {
  const persistentStateReference = useRef<PersistentTableState>({
    sorting: initialSorting,
    columnFilters: initialFilters,
    columnVisibility: {},
    pagination: {
      pageIndex: DEFAULT_PAGE_INDEX,
      pageSize: rowPerPage || DEFAULT_PAGE_SIZE,
    },
  });
  const [sorting, setSorting] = useState<SortingState>(
    persistentStateReference.current.sorting,
  );
  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>(
    persistentStateReference.current.columnFilters,
  );
  const [columnVisibility, setColumnVisibility] = useState<VisibilityState>(
    persistentStateReference.current.columnVisibility,
  );
  const [rowSelection, setRowSelection] = useState({});
  const [pagination, setPagination] = useState<PaginationState>(
    persistentStateReference.current.pagination,
  );

  const handleColumnFilterChange = (event_: Updater<ColumnFiltersState>) => {
    const updatedColumnFilter =
      typeof event_ === "function" ? event_(columnFilters) : event_;

    if (!Array.isArray(updatedColumnFilter)) {
      return [];
    }

    const updatedFilters = updatedColumnFilter.map((filter) => {
      const column = table.getColumn(filter.id);

      return {
        ...filter,
        filterFn: column?.columnDef.meta?.serverFilterFn,
      };
    });

    setColumnFilters(updatedFilters);
    table.setPageIndex(0);
  };

  const handleResetState = () => {
    setSorting(initialSorting);
    setColumnFilters(initialFilters);
    setColumnVisibility({});
    setPagination({
      pageIndex: DEFAULT_PAGE_INDEX,
      pageSize: rowPerPage || DEFAULT_PAGE_SIZE,
    });
  };

  const parsedColumns = useMemo(() => {
    let parsedColumns: ColumnDef<TData, unknown>[] = getParsedColumns({
      visibleColumns,
      columns,
    });

    if (enableRowSelection) {
      parsedColumns = [
        {
          id: "select",
          header: ({ table }) => (
            <Checkbox
              checked={table.getIsAllPageRowsSelected()}
              onChange={() =>
                table.toggleAllPageRowsSelected(
                  !table.getIsAllPageRowsSelected(),
                )
              }
              aria-label="Select all"
            />
          ),
          cell: ({ row }) => (
            <Checkbox
              checked={row.getIsSelected()}
              onChange={() => row.toggleSelected(!row.getIsSelected())}
              aria-label="Select row"
            />
          ),
          enableSorting: false,
          enableHiding: false,
          enableColumnFilter: false,
          enableGlobalFilter: false,
          align: "center",
        },
        ...parsedColumns,
      ];
    }

    if (
      (dataActionsMenu && visibleColumns.length === 0) ||
      (dataActionsMenu && visibleColumns.includes("actions"))
    ) {
      const defaultActionColumn: ColumnDef<TData, unknown> = {
        id: "actions",
        header: () => <i className="pi pi-cog"></i>,
        align: "center",
        cell: ({ row: { original } }) => {
          return (
            <DataActionsMenu
              {...(typeof dataActionsMenu === "function"
                ? dataActionsMenu(original)
                : dataActionsMenu)}
              data={original}
            />
          );
        },
      };

      parsedColumns = [...parsedColumns, defaultActionColumn];
    }

    parsedColumns.forEach((column) => {
      if (column.meta?.filterVariant === "multiselect") {
        column.filterFn = (row, columnId, filterValue) => {
          if (!filterValue || filterValue.length === 0) {
            return row;
          }
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          return filterValue.some((value: any) =>
            row.getValue<unknown[]>(columnId)?.includes(value),
          );
        };
      }
    });

    return parsedColumns;
  }, [visibleColumns, columns]);

  const table = useReactTable({
    data,
    columns: parsedColumns,
    getCoreRowModel: getCoreRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    getPaginationRowModel: getPaginationRowModel(),
    getSortedRowModel: getSortedRowModel(),
    onColumnFiltersChange: handleColumnFilterChange,
    onColumnVisibilityChange: setColumnVisibility,
    onPaginationChange: setPagination,
    onRowSelectionChange: setRowSelection,
    onSortingChange: setSorting,
    state: {
      columnFilters,
      columnVisibility,
      pagination,
      rowSelection,
      sorting,
    },
    manualFiltering: !!fetchData,
    manualSorting: !!fetchData,
    manualPagination: !!fetchData,
    ...(!!fetchData && {
      pageCount: Math.ceil(totalRecords / pagination.pageSize),
    }),
    ...tableOptions,
  });

  const mappedSelectedRows = table.getFilteredSelectedRowModel();
  const totalItems = fetchData
    ? totalRecords
    : table.getFilteredRowModel().rows?.length;
  const storage = useMemo(
    () => getStorage(persistStateStorage),
    [persistStateStorage],
  );

  useEffect(() => {
    // client side rendering with no pagination
    if (!fetchData && !paginated) {
      setPagination((previous) => ({
        ...previous,
        pageSize: data.length,
      }));
    }
  }, [fetchData, data, paginated]);

  useEffect(() => {
    onRowSelectChange && onRowSelectChange(table);
  }, [mappedSelectedRows]);

  useEffect(() => {
    const requestJSON = getRequestJSON(sorting, columnFilters, {
      pageIndex: pagination.pageIndex,
      pageSize: pagination.pageSize,
    });

    fetchData && fetchData(requestJSON);
  }, [
    columnFilters,
    pagination.pageIndex,
    pagination.pageSize,
    sorting,
    fetchData,
  ]);

  useEffect(() => {
    if (visibleColumns.length !== 0) {
      table.setColumnOrder(["select", ...visibleColumns]);
    }
  }, [visibleColumns, parsedColumns]);

  useEffect(() => {
    // Records the table state change without any rerender
    if (persistState && id) {
      persistentStateReference.current = {
        sorting,
        columnFilters,
        columnVisibility,
        pagination,
      };
    }
  }, [id, pagination, persistState, sorting, columnFilters, columnVisibility]);

  useEffect(() => {
    // Restore the table state from storage
    if (!persistState || !id) {
      return;
    }

    const savedState = getSavedTableState(id, storage);

    if (savedState) {
      const { columnFilters, columnVisibility, pagination, sorting } =
        savedState;

      setColumnFilters(columnFilters);
      setColumnVisibility(columnVisibility);
      setSorting(sorting);

      if (rowPerPageOptions.includes(pagination.pageSize)) {
        setPagination(pagination);
      }
    }

    return () => {
      saveTableState(id, persistentStateReference.current, storage);
    };
  }, [id, persistState]);

  useEffect(() => {
    // Store the table state in storage on page refresh
    const handleBeforeUnload = () => {
      if (persistState && id) {
        saveTableState(id, persistentStateReference.current, storage);
      }
    };

    window.addEventListener("beforeunload", handleBeforeUnload);

    return () => {
      window.removeEventListener("beforeunload", handleBeforeUnload);
    };
  }, [id, persistState]);

  return (
    <div id={id} className={("dz-table-container " + className).trimEnd()}>
      {title ? (
        <span children={title.text} data-align={title.align || "center"} />
      ) : null}

      {showColumnsAction || renderToolbarItems || showResetStateAction ? (
        <TableToolbar
          table={table}
          showColumnsAction={showColumnsAction}
          columnActionButtonLabel={columnActionButtonLabel}
          dataActionsMenu={dataActionsMenu}
          showResetStateAction={showResetStateAction}
          renderToolbarItems={renderToolbarItems}
          resetActionButtonLabel={resetStateActionButtonLabel}
          handleResetState={handleResetState}
          enableRowSelection={enableRowSelection}
        />
      ) : null}

      <Table>
        <TableHeader
          table={table}
          renderSortIcons={renderSortIcons}
          inputDebounceTime={inputDebounceTime}
        />

        <TableBody
          table={table}
          parsedColumnsLength={parsedColumns.length}
          emptyTableMessage={emptyTableMessage}
          enableRowSelection={enableRowSelection}
          customFormatters={customFormatters}
        />

        {renderTableFooterContent ? (
          <TableFooter children={renderTableFooterContent(table)} />
        ) : null}
      </Table>

      {(!!fetchData || paginated) && totalItems > 0 ? (
        <>
          {renderCustomPagination ? (
            renderCustomPagination(table)
          ) : (
            <Pagination
              currentPage={pagination.pageIndex}
              defaultItemsPerPage={pagination.pageSize}
              onPageChange={(currentPage) => {
                table.setPageIndex(currentPage);
              }}
              itemsPerPageOptions={rowPerPageOptions}
              onItemsPerPageChange={(itemsPerPage) => {
                table.setPageSize(itemsPerPage);
              }}
              totalItems={totalItems}
              {...paginationOptions}
            ></Pagination>
          )}
        </>
      ) : null}

      {isLoading ? (
        <div className="loading-overlay">
          <LoadingIcon color="#55575f" fontSize={"0.5rem"}></LoadingIcon>
        </div>
      ) : null}
    </div>
  );
};

export { DataTable };
