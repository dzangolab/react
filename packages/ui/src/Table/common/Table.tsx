import {
  flexRender,
  getCoreRowModel,
  useReactTable,
  getPaginationRowModel,
  SortingState,
  getSortedRowModel,
  ColumnFiltersState,
  getFilteredRowModel,
  VisibilityState,
  PaginationState,
  Updater,
} from "@tanstack/react-table";
import { Button } from "primereact/button";
import React, {
  SyntheticEvent,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";

import {
  DEFAULT_PAGE_INDEX,
  DEFAULT_PAGE_PER_OPTIONS,
  DEFAULT_PAGE_SIZE,
} from "./constants";
import {
  Table,
  TableBody,
  TableCell,
  ColumnHeader,
  TableHeader,
  TableRow,
  TableToolbar,
  TableFooter,
  TooltipWrapper,
} from "./TableElements";
import { getRequestJSON, getParsedColumns } from "./utils";
import { Checkbox, DebouncedInput, Popup, SortableList } from "../../";
import LoadingIcon from "../../LoadingIcon";
import { Pagination } from "../../Pagination";

import type { TCustomColumnFilter, TDataTableProperties } from "./types";
import type { Cell, Column, ColumnDef } from "@tanstack/react-table";

const DataTable = <TData extends { id: string | number }>({
  border = "grid",
  className = "",
  columns = [],
  columnActionBtnLabel: columnActionButtonLabel = "Columns",
  data,
  emptyTableMessage = "No results.",
  enableRowSelection = false,
  id,
  isLoading = false,
  initialFilters = [],
  inputDebounceTime,
  fetchData,
  renderToolbarItems,
  renderTableFooterContent,
  renderCustomPagination,
  renderSortIcons,
  title,
  paginated = true,
  rowPerPage,
  rowPerPageOptions = DEFAULT_PAGE_PER_OPTIONS,
  visibleColumns = [],
  onRowSelectChange,
  totalRecords = 0,
  paginationOptions,
  stripe = "none",
  showColumnsAction = false,
  ...tableOptions
}: TDataTableProperties<TData>) => {
  const [sorting, setSorting] = useState<SortingState>([]);
  const [columnFilters, setColumnFilters] = useState<TCustomColumnFilter[]>(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    initialFilters as any,
  );
  const [columnVisibility, setColumnVisibility] = useState<VisibilityState>({});
  const [rowSelection, setRowSelection] = useState({});
  const [pagination, setPagination] = useState<PaginationState>({
    pageIndex: DEFAULT_PAGE_INDEX,
    pageSize: rowPerPage || DEFAULT_PAGE_SIZE,
  });
  const [isFilterRowVisible, setIsFilterRowVisible] = useState(false);

  const handleColumnFilterChange = (event_: Updater<ColumnFiltersState>) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setColumnFilters(event_ as any);
  };

  const parsedColumns = useMemo(
    () => getParsedColumns({ visibleColumns, columns }),
    [visibleColumns, columns],
  );

  const columnsWithRowSelection = useMemo(() => {
    if (!enableRowSelection) {
      return parsedColumns;
    }

    const columns: ColumnDef<TData>[] = [
      {
        id: "select",
        header: ({ table }) => (
          <Checkbox
            checked={table.getIsAllPageRowsSelected()}
            onChange={() =>
              table.toggleAllPageRowsSelected(!table.getIsAllPageRowsSelected())
            }
            aria-label="Select all"
          />
        ),
        cell: ({ row }) => (
          <Checkbox
            checked={row.getIsSelected()}
            onChange={() => row.toggleSelected(!row.getIsSelected())}
            aria-label="Select row"
          />
        ),
        enableSorting: false,
        enableHiding: false,
        enableColumnFilter: false,
        enableGlobalFilter: false,
        align: "center",
      },
      ...parsedColumns,
    ];

    return columns;
  }, [parsedColumns]);

  const table = useReactTable({
    data,
    columns: columnsWithRowSelection,
    getCoreRowModel: getCoreRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    getPaginationRowModel: getPaginationRowModel(),
    getSortedRowModel: getSortedRowModel(),
    onColumnFiltersChange: handleColumnFilterChange,
    onColumnVisibilityChange: setColumnVisibility,
    onPaginationChange: setPagination,
    onRowSelectionChange: setRowSelection,
    onSortingChange: setSorting,
    state: {
      columnFilters,
      columnVisibility,
      pagination,
      rowSelection,
      sorting,
    },
    manualFiltering: !!fetchData,
    manualSorting: !!fetchData,
    manualPagination: !!fetchData,
    ...(!!fetchData && {
      pageCount: Math.ceil(totalRecords / pagination.pageSize),
    }),
    ...tableOptions,
  });

  const mappedSelectedRows = table.getFilteredSelectedRowModel();

  useEffect(() => {
    onRowSelectChange && onRowSelectChange(table);
  }, [mappedSelectedRows]);

  useEffect(() => {
    const requestJSON = getRequestJSON(sorting, columnFilters, {
      pageIndex: pagination.pageIndex,
      pageSize: pagination.pageSize,
    });

    fetchData && fetchData(requestJSON);
  }, [columnFilters, pagination.pageIndex, pagination.pageSize, sorting]);

  const handleSort = useCallback(
    (event: SyntheticEvent, sortHandler?: (event: SyntheticEvent) => void) => {
      event.stopPropagation();
      if (sortHandler) {
        sortHandler(event);
      }
    },
    [],
  );

  useEffect(() => {
    if (visibleColumns.length !== 0) {
      table.setColumnOrder(["select", ...visibleColumns]);
    }
  }, [visibleColumns, columnsWithRowSelection]);

  const renderTooltipContent = (
    cell: Cell<TData, unknown>,
  ): React.ReactNode => {
    if (typeof cell.column.columnDef.tooltip === "string") {
      return cell.column.columnDef.tooltip;
    } else if (typeof cell.column.columnDef.tooltip === "function") {
      return cell.column.columnDef.tooltip(cell);
    }

    return cell.getValue() as string;
  };

  const totalItems = fetchData
    ? totalRecords
    : table.getFilteredRowModel().rows?.length;

  return (
    <div
      id={id}
      data-border={border}
      className={"table-container " + className}
    >
      {title ? (
        <h6
          className="title"
          children={title.text}
          data-align={title.align || "center"}
        />
      ) : null}

      {showColumnsAction || renderToolbarItems ? (
        <TableToolbar
          children={
            <>
              {showColumnsAction ? (
                <Popup
                  trigger={<Button label={columnActionButtonLabel} />}
                  content={
                    <SortableList
                      items={table
                        .getAllLeafColumns()
                        .filter((column) => column.id !== "select")
                        .map((column, index) => ({
                          id: index,
                          data: column,
                          // eslint-disable-next-line @typescript-eslint/no-explicit-any
                          render: (data: any) => {
                            let header = data.columnDef.header;

                            if (typeof data.columnDef.header === "function") {
                              header = data.columnDef.header();
                            }

                            return (
                              <>
                                <Checkbox
                                  checked={data.getIsVisible()}
                                  onClick={() => data.toggleVisibility()}
                                  label={header}
                                ></Checkbox>
                              </>
                            );
                          },
                        }))}
                      onSort={(sorted) => {
                        table.setColumnOrder([
                          ...(enableRowSelection ? ["select"] : []),
                          ...sorted.map((item) => item.data.id),
                        ]);
                      }}
                    />
                  }
                />
              ) : null}

              {renderToolbarItems ? renderToolbarItems(table) : null}
            </>
          }
        />
      ) : null}

      <Table data-stripe={stripe}>
        <TableHeader>
          {table.getHeaderGroups().map((headerGroup) => (
            <TableRow key={headerGroup.id} className="header-row">
              {headerGroup.headers.map(
                ({ column, getContext, id, isPlaceholder, colSpan }) => {
                  const {
                    columnDef,
                    getCanSort,
                    getIsSorted,
                    getToggleSortingHandler,
                  } = column;

                  const activeColumnClass = `${
                    getIsSorted() === "asc" || getIsSorted() === "desc"
                      ? "highlight"
                      : ""
                  }`;

                  if (!isFilterRowVisible && column.getCanFilter()) {
                    setIsFilterRowVisible(true);
                  }

                  const getSortIcon = () => {
                    if (renderSortIcons) {
                      return renderSortIcons(getIsSorted());
                    }

                    switch (getIsSorted()) {
                      case "asc":
                        return <i className="pi pi-arrow-up"></i>;
                      case "desc":
                        return <i className="pi pi-arrow-down"></i>;
                      default:
                        return <i className="pi pi-arrows-v"></i>;
                    }
                  };

                  return (
                    <ColumnHeader
                      key={id}
                      colSpan={colSpan}
                      className={`column-${id} ${activeColumnClass}`}
                      data-align={columnDef.align || "left"}
                      onClick={(event) => {
                        if (getCanSort()) {
                          handleSort(event, getToggleSortingHandler());
                        }
                      }}
                    >
                      <>
                        {isPlaceholder ? null : (
                          <>
                            {flexRender(columnDef.header, getContext())}

                            {getCanSort() ? (
                              <span className="sort-state">
                                {getSortIcon()}
                              </span>
                            ) : null}
                          </>
                        )}
                      </>
                    </ColumnHeader>
                  );
                },
              )}
            </TableRow>
          ))}

          {isFilterRowVisible ? (
            <TableRow key={"filters"} className={`header-row filters`}>
              {table.getVisibleLeafColumns().map((column) => {
                if (!column.getCanFilter()) {
                  return <TableCell key={"filter" + column.id}></TableCell>;
                }

                const activeColumnClass = `${
                  column.getIsFiltered() ? "highlight" : ""
                }`;

                return (
                  <ColumnHeader
                    key={"filter" + column.id}
                    data-label={column.id}
                    data-align={column.columnDef.align || "left"}
                    className={
                      (column.id ? `column-${column.id} ` : ``) +
                      activeColumnClass
                    }
                  >
                    <DebouncedInput
                      defaultValue={column.getFilterValue() as string}
                      onInputChange={(value) => {
                        column.setFilterValue(value);
                      }}
                      placeholder={column.columnDef.filterPlaceholder || ""}
                      debounceTime={inputDebounceTime}
                    ></DebouncedInput>
                  </ColumnHeader>
                );
              })}
            </TableRow>
          ) : null}
        </TableHeader>

        <TableBody>
          {table.getRowModel().rows?.length ? (
            table.getRowModel().rows.map((row) => (
              <TableRow
                key={row.id}
                data-state={row.getIsSelected() && "selected"}
                data-id={row.original.id ?? row.id}
              >
                {row.getVisibleCells().map((cell) => {
                  return (
                    <TableCell
                      key={cell.id}
                      data-label={cell.column.id}
                      data-align={cell.column.columnDef.align || "left"}
                      className={
                        cell.column.id ? `column-${cell.column.id}` : ``
                      }
                    >
                      {cell.column.columnDef.tooltip ? (
                        <TooltipWrapper
                          tooltipOptions={{
                            children: renderTooltipContent(cell),
                            ...cell.column.columnDef?.tooltipOptions,
                          }}
                          cellContent={flexRender(
                            cell.column.columnDef.cell,
                            cell.getContext(),
                          )}
                        ></TooltipWrapper>
                      ) : (
                        flexRender(
                          cell.column.columnDef.cell,
                          cell.getContext(),
                        )
                      )}
                    </TableCell>
                  );
                })}
              </TableRow>
            ))
          ) : (
            <TableRow>
              <TableCell colSpan={columnsWithRowSelection.length}>
                {emptyTableMessage}
              </TableCell>
            </TableRow>
          )}
        </TableBody>

        {renderTableFooterContent ? (
          <TableFooter children={renderTableFooterContent(table)} />
        ) : null}
      </Table>

      {paginated && totalItems > 0 ? (
        <>
          {renderCustomPagination ? (
            renderCustomPagination(table)
          ) : (
            <Pagination
              currentPage={pagination.pageIndex}
              defaultItemsPerPage={pagination.pageSize}
              onPageChange={(currentPage) => {
                table.setPageIndex(currentPage);
              }}
              itemsPerPageOptions={rowPerPageOptions}
              onItemsPerPageChange={(itemsPerPage) => {
                table.setPageSize(itemsPerPage);
              }}
              totalItems={totalItems}
              {...paginationOptions}
            ></Pagination>
          )}
        </>
      ) : null}

      {isLoading ? (
        <div className="loading-overlay">
          <LoadingIcon color="#55575f" fontSize={"0.5rem"}></LoadingIcon>
        </div>
      ) : null}
    </div>
  );
};

export { DataTable };
