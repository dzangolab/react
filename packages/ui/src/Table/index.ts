import "./table.css";

import BaseTable from "./common/BaseTable";
import { TableContext } from "./common/TableContext";
import TableDetail from "./common/TableDetail";
import TableFooter from "./common/TableFooter";
import TableProvider from "./TableProvider";

export {
  BaseTable,
  TableContext,
  TableDetail,
  TableFooter,
  TableProvider as Table,
};

export type { TableColumnDefinition, TRequestJSON } from "./common/types";
