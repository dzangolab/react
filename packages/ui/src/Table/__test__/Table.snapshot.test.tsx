import React from "react";
import renderer, { create } from "react-test-renderer";
import { expect, test } from "vitest";

import Table from "../TableProvider";

const columns = [
  {
    accessorKey: "id",
    header: () => <span>ID</span>,
  },
  {
    accessorKey: "name",
    header: () => <span>Name</span>,
  },
];

const data = [
  { id: 1, name: "jone" },
  { id: 2, name: "abhi" },
  { id: 3, name: "mike" },
  { id: 4, name: "tony" },
];

const fetcher = () => {
  return data;
};

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = create(
    <Table
      columns={columns}
      data={data}
      fetcher={fetcher}
      totalItems={data.length}
    />
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
