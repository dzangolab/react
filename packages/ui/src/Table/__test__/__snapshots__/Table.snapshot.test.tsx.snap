// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<div
  className="sticky-header hide-scrollbar"
>
  <div
    className="table-wrapper"
  >
    <table>
      <thead>
        <tr
          className="table-header-1"
        >
          <th
            colSpan={1}
            style={
              {
                "width": 150,
              }
            }
          >
            <div
              className="disable-select"
              onClick={[Function]}
            >
              <span>
                ID
              </span>
              <button
                className="sort-button"
              >
                ‚ÜïÔ∏è
              </button>
            </div>
          </th>
          <th
            colSpan={1}
            style={
              {
                "width": 150,
              }
            }
          >
            <div
              className="disable-select"
              onClick={[Function]}
            >
              <span>
                Name
              </span>
              <button
                className="sort-button"
              >
                ‚ÜïÔ∏è
              </button>
              <div
                className="filter-wrapper"
                onClick={[Function]}
              >
                <button
                  onClick={[Function]}
                >
                  üôà
                </button>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            style={
              {
                "width": 150,
              }
            }
          >
            1
          </td>
          <td
            style={
              {
                "width": 150,
              }
            }
          >
            jone
          </td>
        </tr>
        <tr>
          <td
            style={
              {
                "width": 150,
              }
            }
          >
            2
          </td>
          <td
            style={
              {
                "width": 150,
              }
            }
          >
            abhi
          </td>
        </tr>
        <tr>
          <td
            style={
              {
                "width": 150,
              }
            }
          >
            3
          </td>
          <td
            style={
              {
                "width": 150,
              }
            }
          >
            mike
          </td>
        </tr>
        <tr>
          <td
            style={
              {
                "width": 150,
              }
            }
          >
            4
          </td>
          <td
            style={
              {
                "width": 150,
              }
            }
          >
            tony
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="table-info-container"
  >
    <div
      className="page-details"
    >
      <span
        className="current-page-info"
      >
        <strong>
          1 of 1 
        </strong>
        |
      </span>
      <div
        className="page-controller"
      >
        <span>
           Go to page:
        </span>
        <input
          disabled={true}
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          size={1}
          type="number"
          value={1}
        />
        <select
          onChange={[Function]}
          value={10}
        >
          <option
            value={10}
          >
            Show 
            10
          </option>
          <option
            value={20}
          >
            Show 
            20
          </option>
          <option
            value={30}
          >
            Show 
            30
          </option>
        </select>
      </div>
    </div>
    <div
      className="pagination-container"
    >
      <button
        className="pagination-button"
        disabled={true}
        onClick={[Function]}
      >
        &lt;&lt;
      </button>
      <button
        className="pagination-button"
        disabled={true}
        onClick={[Function]}
      >
        &lt;
      </button>
      <button
        className="pagination-button"
        disabled={true}
        onClick={[Function]}
      >
        &gt;
      </button>
      <button
        className="pagination-button"
        disabled={true}
        onClick={[Function]}
      >
        &gt;&gt;
      </button>
    </div>
  </div>
</div>
`;
